### PR TITLE
[JENKINS-69032] Plugin manager update site URL can be saved empty

### PR DESCRIPTION
**Pull request by benebsiny**
See [JENKINS-69032](https://issues.jenkins.io/browse/JENKINS-69032).

update-site

Form validation and "Reset to default" button are implemented.

### Proposed changelog entries

* Add a "Reset to default" button to reset update site url to default.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6886"></a>

### CODE REVIEW

**Review by NotMyFault — COMMENTED** [review]
Does the FormValidation validate the input though? It looks like you'd be able to hit the submission button still, even if the field is empty.

**Comment by NotMyFault** [comment]
> The FormValidation can only shows the error message at most

Yes, but the change proposed doesn't address the issue linked, it just adds a warning and a button to reset the field.

**Comment by daniel-beck** [comment]
Given that accepting user submissions even if they're nonsensical is a fairly common pattern in Jenkins, validation + reset button doesn't seem like a bad solution.

**Review by NotMyFault — APPROVED** [review]

**Comment by daniel-beck** [comment]
It might also be useful to review the error that appears when you check for updates from an empty URL update site to make sure that explains the configuration is screwed up and should be changed. Not necessary if you don't feel like adding that though.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (timja) [inline core/src/main/resources/hudson/PluginManager/advanced.jelly:85]: this would introduce new CSP errors from my understanding.

I think you need to do something like:

```suggestion
                                 checkUrl="'checkUpdateSiteUrl" checkDependsOn=""/>
```

@daniel-beck can you check this please?

Comment (daniel-beck) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:85]: Good point, thanks for spotting it 👍  From my understanding this should work. Manual test should confirm.

See also https://github.com/jenkins-infra/jenkins.io/pull/5301/files#diff-4405e13cac1124b62201d74e661384a7eeccc56aa8515e5678bd97503631a5a9R70-R76

Comment (benebsiny) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly]: Thanks for your suggestion!
Notice that Deploy Plugin send the request in the same way. It may be addressed by opening another issue.

Comment (daniel-beck) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly]: >  Notice that Deploy Plugin send the request in the same way. It may be addressed by opening another issue.

Right, this is a fairly recent project without published docs, JEP, or other guidelines :) Once we know how plugin maintainers should implement such things and have it documented, we can make a case for plugins to be improved.

File: core/src/main/resources/hudson/PluginManager/_updateSite.js
Comment (jtnord) [inline core/src/main/resources/hudson/PluginManager/_updateSite.js:4]: This may or may not be a default. so assuming it is the default would be wrong.

There may be one or more sites configured, this is called out in the Javadoc and installed system properties and/or plugins can (and do) remove / change the default UpdateSite as an when they are started.
This would break the places where it happens.

if this is to be included it should only be performed if there is a [single site](https://javadoc.jenkins.io/hudson/model/UpdateCenter.html#getSites()) defined and its [id](https://javadoc.jenkins.io/hudson/model/UpdateSite.html#getId()) is the generic [predefined one](https://javadoc.jenkins.io/hudson/model/UpdateCenter.html#PREDEFINED_UPDATE_SITE_ID).

**Review by jtnord — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/PluginManager.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginManager.java:1894]: rather than just checking the URL can be parsed should a request be made to see if there is even something serving JSON (html with embedded json 😱 ) at what would be given constructed URL?

Comment (daniel-beck) [reply core/src/main/java/hudson/PluginManager.java:1894]: Oooh, great idea. (Probably needs permission check unless `PluginManager` is entirely protected.)

Comment (daniel-beck) [reply core/src/main/java/hudson/PluginManager.java:1894]: Also needs to be mindful of the changes done to the URL (adding query parameters).

File: core/src/main/resources/hudson/PluginManager/_updateSite.js
Comment (benebsiny) [reply core/src/main/resources/hudson/PluginManager/_updateSite.js:4]: The predefined value will be overwritten when a user sets a new URL. It seems that a new API needs be developed. (But not sure if it's possible to get the default one after it is overwritten)

Comment (jtnord) [reply core/src/main/resources/hudson/PluginManager/_updateSite.js:4]: or you could only display the `reset to default` if the above conditions are met (no new APIs are needed to see if this is the case, but you may need something to expose to the jelly to conditionally show it)

Comment (benebsiny) [reply core/src/main/resources/hudson/PluginManager/_updateSite.js:4]: @jtnord  I have made some changes. Now the button will be shown if there's only one site configured.

File: core/src/main/java/hudson/PluginManager.java
Comment (benebsiny) [reply core/src/main/java/hudson/PluginManager.java:1894]: @jtnord Update Site also accepts JSONP format. It seems that there's no method to validate the JSONP format😅.

Comment (jtnord) [reply core/src/main/java/hudson/PluginManager.java:1894]: a simple `HEAD` for the url with the version parameter appended (`?version=jenkins.version`) and check that you get a valid (non error) HTTP response rather than a timeout would likely catch a number of typos.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (jtnord) [inline core/src/main/resources/hudson/PluginManager/advanced.jelly:81]: please only enable this if **all** of the following are true:

1. there exists an `UpdateSite` with an `id` equal to [`hudson.model.UpdateCenter#PREDEFINED_UPDATE_SITE_ID`](https://github.com/jenkinsci/jenkins/blob/46422c0a54371738e395ab49c11d0501527231e4/core/src/main/java/hudson/model/UpdateCenter.java#L166)
2. [`hudson.model.UpdateCenter#ID_DEFAULT`](https://github.com/jenkinsci/jenkins/blob/46422c0a54371738e395ab49c11d0501527231e4/core/src/main/java/hudson/model/UpdateCenter.java#L172) is equal to  [`hudson.model.UpdateCenter#PREDEFINED_UPDATE_SITE_ID`](https://github.com/jenkinsci/jenkins/blob/46422c0a54371738e395ab49c11d0501527231e4/core/src/main/java/hudson/model/UpdateCenter.java#L166)

if either of the above are not true the update site lists have been customized and then the "default" that is being reset to will in most cases not infact be the correct value.

**Review by jtnord — CHANGES_REQUESTED** [review]

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (jtnord) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:81]: actually it is only number 2.  because the save would create the update site with id [`ID_DEFAULT`](https://github.com/jenkinsci/jenkins/blob/46422c0a54371738e395ab49c11d0501527231e4/core/src/main/java/hudson/model/UpdateCenter.java#L166) if not present.

File: core/src/main/java/hudson/PluginManager.java
Comment (benebsiny) [reply core/src/main/java/hudson/PluginManager.java]: @jtnord Now the method will send the requests and check if the connection failed (exception thrown).

Comment (daniel-beck) [reply core/src/main/java/hudson/PluginManager.java]: Note that the connection test feature adds the `uctest` query parameter. This serves bandwidth for the infra, because we can send an empty response: https://github.com/jenkinsci/jenkins/blob/237b9ca457c032885ed98e29620f0ed8c0e55ca0/core/src/main/java/hudson/model/UpdateCenter.java#L1208

File: core/src/main/java/hudson/PluginManager.java
Comment (basil) [inline core/src/main/java/hudson/PluginManager.java:1936]: This method does not properly handle proxy servers. Per its Javadoc, `hudson.ProxyConfiguration#open` should be used wherever `URL#openConnection()` to internet URLs is invoked directly. Please update the code accordingly and describe your testing against a proxy server. To test, I would recommend finding a Docker image that provides a proxy server and starting up a container from that image.

Comment (benebsiny) [reply core/src/main/java/hudson/PluginManager.java]: Thanks for the suggestion! I've tested it with docker container [datadog/squid](https://hub.docker.com/r/datadog/squid). After setting the proxy in the `HTTP Proxy Configuration` section, the `hudson.ProxyConfiguration#open` works fine to check the connectivity. Have a look at the new commit.

File: core/src/main/java/hudson/PluginManager.java
Comment (benebsiny) [reply core/src/main/java/hudson/PluginManager.java]: @daniel-beck The parameter is added. Thanks for the suggestion!

File: core/src/main/java/hudson/PluginManager.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginManager.java:1928]: This needs to be protected by a check to ensure the user has admin permissions to prevent server side request forgery.

```suggestion
    @RequirePOST public FormValidation doCheckUpdateSiteUrl(StaplerRequest request, @QueryParameter String value) {
        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
```

**Review by jtnord — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/PluginManager.java
Comment (benebsiny) [reply core/src/main/java/hudson/PluginManager.java:1928]: Yess, this should be protected. Thanks for the suggestion!

**Review by jtnord — APPROVED** [review]
Looks ok to me.  Chinese translation passed the google translate check

File: core/src/main/java/hudson/PluginManager.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginManager.java:1939]: should this also have a read timeout?

```suggestion
                conn.setReadTimeout(1000);
```

File: core/src/main/java/hudson/PluginManager.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginManager.java:1943]: worth logging to help diagnosis? (untested)

```suggestion
                    LOGGER.log(Level.FINE, "Obtained a non OK ({0})response from the update center, new Object[]{conn.getResponseCode(), url});
                    return FormValidation.error(Messages.PluginManager_connectionFailed());
```

File: core/src/main/java/hudson/PluginManager.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginManager.java:1946]: worth logging to help diagnosis? (untested)

```suggestion
                LOGGER.log(Level.FINE, "failed to check update site", e );
                return FormValidation.error(Messages.PluginManager_connectionFailed());
```

**Review by jtnord — APPROVED** [review]
LGTM - left a few untested suggestions.

File: core/src/main/java/hudson/PluginManager.java
Comment (benebsiny) [reply core/src/main/java/hudson/PluginManager.java:1939]: Will 1 sec too short? How about 5 sec which is same as connect timeout?

File: core/src/main/java/hudson/PluginManager.java
Comment (benebsiny) [reply core/src/main/java/hudson/PluginManager.java:1945]: Tested, this looks good.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (benebsiny) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly]: Fixed. The result now looks like this.

record

**Review by NotMyFault — APPROVED** [review]
Thanks for addressing the outstanding concerns, looks promising so far.

File: core/src/main/resources/hudson/PluginManager/_updateSite.js
Comment (timja) [inline core/src/main/resources/hudson/PluginManager/_updateSite.js:22]: the UX isn't very great here.

If I have this:
image

then I make a change
image

I click the Submit button:
image

Instead of the button being submitted I get the `Reset to default` appearing and my change isn't saved.

I this this should only be shown on page load, and only disappear if it's clicked, doesn't need to be as 'smart' as it is

**Review by timja — CHANGES_REQUESTED** [review]

File: core/src/main/resources/hudson/PluginManager/_updateSite.js
Comment (benebsiny) [reply core/src/main/resources/hudson/PluginManager/_updateSite.js:15]: @timja Thanks for the suggestion! Have a look at the latest committed changes.

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by jtnord** [comment]
caused https://issues.jenkins.io/browse/JENKINS-70240

**Pull request merged**
